internal/controller/controllers/kube_compat.go

```diff
@@ -2,19 +2,195 @@ package controllers
+	}
+	pathTypePrefix := netv1.PathTypePrefix
+	ingress := &netv1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: asc.namespace,
+		},
+	}
+
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(asc.Object, ingress, asc.rec.Scheme); err != nil {
+			return err
+		}
+		ingress.Spec = netv1.IngressSpec{
+			IngressClassName: asc.spec.Ingress.ClassName,
+			Rules: []netv1.IngressRule{{
+				Host: host,
+				IngressRuleValue: netv1.IngressRuleValue{HTTP: &netv1.HTTPIngressRuleValue{
+					Paths: []netv1.HTTPIngressPath{{
+						Path:     "/",
+						PathType: &pathTypePrefix,
+						Backend: netv1.IngressBackend{
+							Service: &netv1.IngressServiceBackend{
+								Name: name,
+								Port: netv1.ServiceBackendPort{
+									Number: port,
+								},
+							},
+						},
+					}},
+				}},
+			}},
+		}
+		return nil
 	}
-	return false, client.IgnoreNotFound(err)
+
+	return ingress, mutateFn, nil
+}
+
+func certManagerComponents() []component {
+	return []component{
+		{"SelfSignedIssuer", aiv1beta1.ReasonCertificateFailure, newSelfSignedIssuer},
+		{"CAIssuer", aiv1beta1.ReasonCertificateFailure, newCAIssuer},
+		{"CACert", aiv1beta1.ReasonCertificateFailure, newCACert},
+		{"WebhookCert", aiv1beta1.ReasonCertificateFailure, newWebhookCert},
+	}
+}
+
+// newSelfSignedIssuer describes how to create an Issuer for the self-signed root issuer
+func newSelfSignedIssuer(ctx context.Context, log logrus.FieldLogger, asc ASC) (client.Object, controllerutil.MutateFn, error) {
+	is := &certtypes.Issuer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      selfSignedIssuerName,
+			Namespace: asc.namespace,
+		},
+	}
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(asc.Object, is, asc.rec.Scheme); err != nil {
+			return err
+		}
+		is.Spec = certtypes.IssuerSpec{
+			IssuerConfig: certtypes.IssuerConfig{
+				SelfSigned: &certtypes.SelfSignedIssuer{},
+			},
+		}
+		return nil
+	}
+	return is, mutateFn, nil
+}
+
+// newCACert describes how to create a Certificate for the root CA
+func newCACert(ctx context.Context, log logrus.FieldLogger, asc ASC) (client.Object, controllerutil.MutateFn, error) {
+	cert := &certtypes.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      caIssuerName,
+			Namespace: asc.namespace,
+		},
+	}
+
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(asc.Object, cert, asc.rec.Scheme); err != nil {
+			return err
+		}
+		cert.Spec.CommonName = caCommonName
+		cert.Spec.IsCA = true
+		cert.Spec.SecretName = caIssuerName
+		cert.Spec.IssuerRef = certmeta.ObjectReference{
+			Kind: "Issuer",
+			Name: selfSignedIssuerName,
+		}
+		return nil
+	}
+
+	return cert, mutateFn, nil
+}
+
+// newCAIssuer describes how to create an Issuer for creating certificates signed by the root CA cert
+func newCAIssuer(ctx context.Context, log logrus.FieldLogger, asc ASC) (client.Object, controllerutil.MutateFn, error) {
```

Comment:
Same here, does `ctx context.Context, log logrus.FieldLogger` need to be in the argument list? :o

Reply:
Yes, because this is a [`NewComponentFn`](https://github.com/openshift/assisted-service/blob/d2bba26c003c543612de88883f936d54aceb4ca2/internal/controller/controllers/agentserviceconfig_controller.go#L187)